pkg/security/admission/admission.go

```diff
@@ -0,0 +1,403 @@
+
+	for _, c := range store.List() {
+		constraint, ok := c.(*kapi.SecurityContextConstraints)
+		if !ok {
+			return nil, errors.NewInternalError(fmt.Errorf("error converting object from store to a security context constraint: %v", c))
+		}
+		if constraintAppliesTo(constraint, userInfo) {
+			matchedConstraints = append(matchedConstraints, constraint)
+		}
+	}
+
+	return matchedConstraints, nil
+}
+
+// constraintAppliesTo inspects the constraint's users and groups against the userInfo to determine
+// if it is usable by the userInfo.
+func constraintAppliesTo(constraint *kapi.SecurityContextConstraints, userInfo user.Info) bool {
+	for _, user := range constraint.Users {
+		if userInfo.GetName() == user {
+			return true
+		}
+	}
+	for _, userGroup := range userInfo.GetGroups() {
+		if constraintSupportsGroup(userGroup, constraint.Groups) {
+			return true
+		}
+	}
+	return false
+}
+
+// constraintSupportsGroup checks that group is in constraintGroups.
+func constraintSupportsGroup(group string, constraintGroups []string) bool {
+	for _, g := range constraintGroups {
+		if g == group {
+			return true
+		}
+	}
+	return false
+}
+
+// getPreallocatedUIDRange retrieves the annotated value from the service account, splits it to make
+// the min/max and formats the data into the necessary types for the strategy options.
+func getPreallocatedUIDRange(ns *kapi.Namespace) (*int64, *int64, error) {
+	annotationVal, ok := ns.Annotations[allocator.UIDRangeAnnotation]
+	if !ok {
+		return nil, nil, fmt.Errorf("unable to find annotation %s", allocator.UIDRangeAnnotation)
+	}
+	if len(annotationVal) == 0 {
+		return nil, nil, fmt.Errorf("found annotation %s but it was empty", allocator.UIDRangeAnnotation)
+	}
+	uidBlock, err := uid.ParseBlock(annotationVal)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var min int64 = int64(uidBlock.Start)
+	var max int64 = int64(uidBlock.End)
+	glog.V(4).Infof("got preallocated values for min: %d, max: %d for uid range in namespace %s", min, max, ns.Name)
+	return &min, &max, nil
+}
+
+// getPreallocatedLevel gets the annotated value from the service account.
+func getPreallocatedLevel(ns *kapi.Namespace) (string, error) {
+	level, ok := ns.Annotations[allocator.MCSAnnotation]
+	if !ok {
+		return "", fmt.Errorf("unable to find annotation %s", allocator.MCSAnnotation)
+	}
+	if len(level) == 0 {
+		return "", fmt.Errorf("found annotation %s but it was empty", allocator.MCSAnnotation)
+	}
+	glog.V(4).Infof("got preallocated value for level: %s for selinux options in namespace %s", level, ns.Name)
+	return level, nil
+}
+
+// requiresPreAllocatedUIDRange returns true if the strategy is must run in range and the min or max
+// is not set.
+func requiresPreAllocatedUIDRange(constraint *kapi.SecurityContextConstraints) bool {
+	return constraint.RunAsUser.Type == kapi.RunAsUserStrategyMustRunAsRange &&
+		(constraint.RunAsUser.UIDRangeMin == nil && constraint.RunAsUser.UIDRangeMax == nil)
```

Comment:
so, if both are nil, we look up the range from the namespace and set both.
if one is nil, we do nothing, and if the min happened to be nil, the uid ends up as nil. Is that working as desired?